vmbindings/dummyvm/src/tests/doc_mutator_storage.rs

```diff
@@ -0,0 +1,140 @@
+}
+
+#[test]
+pub fn boxed_pointer() {
+    MMTK_SINGLETON.with_fixture(|fixture| {
+        let tls_opaque_pointer = VMMutatorThread(VMThread(OpaquePointer::UNINITIALIZED));
+
+        // ANCHOR: mutator_storage_boxed_pointer
+        struct MutatorInTLS {
+            // Store the mutator as a boxed pointer.
+            // Accessing any value in the mutator will need a dereferencing of the boxed pointer.
+            ptr: Box<Mutator<DummyVM>>,
+        }
+
+        // Bind an MMTk mutator
+        let mutator = mmtk::memory_manager::bind_mutator(&fixture.mmtk, tls_opaque_pointer);
+        // Store the pointer in TLS
+        let mut storage = MutatorInTLS { ptr: mutator };
+
+        // Allocate
+        let addr =
+            mmtk::memory_manager::alloc(&mut storage.ptr, 8, 8, 0, AllocationSemantics::Default);
+        // ANCHOR_END: mutator_storage_boxed_pointer
+
+        assert!(!addr.is_zero());
+    });
+}
+
+#[test]
+pub fn embed_mutator_struct() {
+    MMTK_SINGLETON.with_fixture(|fixture| {
+        let tls_opaque_pointer = VMMutatorThread(VMThread(OpaquePointer::UNINITIALIZED));
+
+        // ANCHOR: mutator_storage_embed_mutator_struct
+        struct MutatorInTLS {
+            embed: Mutator<DummyVM>,
+        }
+
+        // Bind an MMTk mutator
+        let mutator = mmtk::memory_manager::bind_mutator(&fixture.mmtk, tls_opaque_pointer);
+        // Store the struct (or use memcpy for non-Rust code)
+        let mut storage = MutatorInTLS { embed: *mutator };
+        // Allocate
+        let addr =
+            mmtk::memory_manager::alloc(&mut storage.embed, 8, 8, 0, AllocationSemantics::Default);
+        // ANCHOR_END: mutator_storage_embed_mutator_struct
+
+        assert!(!addr.is_zero());
+    })
+}
+
+#[test]
+pub fn embed_fastpath_struct() {
+    MMTK_SINGLETON.with_fixture(|fixture| {
+        let tls_opaque_pointer = VMMutatorThread(VMThread(OpaquePointer::UNINITIALIZED));
+
+        // ANCHOR: mutator_storage_embed_fastpath_struct
+        use mmtk::util::alloc::BumpPointer;
+        struct MutatorInTLS {
+            default_bump_pointer: BumpPointer,
+            mutator: Box<Mutator<DummyVM>>,
+        }
+
+        // Bind an MMTk mutator
+        let mutator = mmtk::memory_manager::bind_mutator(&fixture.mmtk, tls_opaque_pointer);
+        // Create a fastpath BumpPointer
+        let default_bump_pointer = {
+            // First find the allocator
+            let selector = mmtk::memory_manager::get_allocator_mapping(
+                &fixture.mmtk,
+                AllocationSemantics::Default,
+            );
+            let default_allocator = unsafe {
+                mutator.allocator_impl::<mmtk::util::alloc::BumpAllocator<DummyVM>>(selector)
+            };
+            // Copy the cursor/limit value to the BumpPointer struct
+            BumpPointer::new(
+                default_allocator.bump_pointer.cursor,
+                default_allocator.bump_pointer.limit,
+            )
+        };
```

Comment:
I was trying to avoid making an assumption about the initial values for a `BumpPointer` at the binding side. So I explicitly get and set the values in the example. Using `Default` sounds better than assuming zero values. But it still assumes 1. our allocators always create a `BumpPointer` with `default()`, and 2. when a binding creates `BumpPointer` at the binding side, they always create it for a new mutator/allocator that hasn't been used. Do you think those assumptions are likely to be true in the future? For example, is it likely that multiple threads share one mutator, in which case, newly created threads may use a `BumpPointer` that has non-zero values.

If we are not certain, I think it is okay to explicitly set the values at initialisation. Let me know what you think.

Reply:
I remember @steveblackburn talked about this in the past.  He mentioned that the allocators are specifically designed this way so that they will work if the TLS is simply zero-initialized.  We can discuss this in our next group meeting to confirm this.

Reply:
On the other hand, is there any issue if we explicitly get and set the values for `BumpPointer`? I can't see any except that it is more verbose.

Reply:
> On the other hand, is there any issue if we explicitly get and set the values for `BumpPointer`? I can't see any except that it is more verbose.

No.  It is still correct, and doesn't cost much because it is only copied once when creating a Mutator struct.  We can even let `BumpPointer` implement `Copy` so that the code can be even more concise.

Reply:
Actually I added `Copy` to `BumpPointer` in this PR. We do not need to create a new `BumpPointer` here. We can just copy from the allocator.